src/messaging/tests/TestReliableMessageProtocol.cpp

```diff
@@ -95,6 +95,10 @@ class MockAppDelegate : public ExchangeDelegate
                            System::PacketBufferHandle && buffer) override
     {
         IsOnMessageReceivedCalled = true;
+        if (mDropAckResponse)
+        {
+            ec->GetReliableMessageContext()->SetAckPending(false);
```

Comment:
It's not clear to me why this is enough.   ReliableMessageContext has three functions keeping track of acks:

1. `HasPeerRequestedAck()`
2. `IsAckPending()`
3. `GetPendingPeerAckId()`

For standalone acks, it looks like `HasAckPending()` is used.  For piggybacked acks, `HasPeerRequestedAck()` is checked.  I guess in this case there is no actual response so the piggybacking does not come in?  But it's still not clear to me why `HasPeerRequestedAck()` even exists and why its use in `ExchangeMessageDispatch::SendMessage` is correct (instead of testing `IsAckPending()` there).

In particular, once `HasPeerRequestedAck()` starts testing true it _never_ goes back to false afaict... That seems pretty dubious.

If none of that has to do with this PR, please just let me know and I'll file a followup.

Reply:
I think a follow up will be useful for `HasPeerRequestedAck`. This test is relying on standalone ack.

My thinking is that, if it's a piggyback ack, the sender of ack will probably request an ack well (for the message on which ack is being piggybacked). So, if the ack is lost, so is the message on which it's piggybacked. And, that might trigger a retransmit from the ack sender. The overall test case might be trickier, but still doable. Need more analysis if the missing piggybacked ack itself can cause this kind of a deadlock.

Reply:
Filed #7339 on my comments above.